x-pack/plugins/painless_lab/public/application/context.tsx

```diff
@@ -8,50 +8,89 @@ import React, { createContext, ReactNode, useState, useContext } from 'react';
 import { HttpSetup } from 'src/core/public';
 
 import { Links } from '../links';
+import { exampleScript, painlessContextOptions } from './constants';
+import { Store, Payload, Validation } from './types';
 
-import { initialState, Store } from './store';
-import { PAINLESS_LAB_KEY } from './constants';
+interface AppContextProviderArgs {
+  children: ReactNode;
+  value: {
+    http: HttpSetup;
+    links: Links;
+  };
+}
 
 interface ContextValue {
-  state: Store;
-  updateState: (nextState: (s: Store) => Partial<Store>) => void;
+  store: Store;
+  updatePayload: (changes: Partial<Payload>) => void;
   services: {
     http: HttpSetup;
   };
   links: Links;
 }
 
+const initialPayload = {
+  context: painlessContextOptions[0].value,
+  code: exampleScript,
+  parameters: `{
+  "string-parameter": "yay",
+  "number-parameter": 1.5,
+  "boolean-parameter": true
+}`,
+  index: 'default-index',
+  document: `{
+  "my-field": "field-value"
+}`,
+  query: '',
+};
+
 const AppContext = createContext<ContextValue>(undefined as any);
 
-interface AppContextProviderArgs {
-  children: ReactNode;
-  value: {
-    http: HttpSetup;
-    links: Links;
+const validatePayload = (payload: Payload): Validation => {
+  const { index } = payload;
+
+  // For now just validate that the user has entered an index.
+  const indexExists = Boolean(index || index.trim());
+
+  return {
+    isValid: indexExists,
+    fields: {
+      index: indexExists,
+    },
   };
-}
+};
 
 export const AppContextProvider = ({
   children,
   value: { http, links },
 }: AppContextProviderArgs) => {
-  const [state, setState] = useState<Store>(() => ({
-    ...initialState,
+  const PAINLESS_LAB_KEY = 'painlessLabState';
+
+  const defaultPayload = {
```

Comment:
This should be inside of a getter like so that we have this pattern: `useState(() => ...)`. In this implementation we are calling `JSON.parse` and `localStorage.getItem` for every state update. This post explains the lazy init idea: https://dmitripavlutin.com/react-usestate-hook-guide/#3-lazy-initialization-of-state

Reply:
Thanks for sharing that link! I was completely overlooking that performance hit and the link opened my eyes. Will fix.